trin-core/src/utils/hash_delay_queue.rs

```diff
@@ -0,0 +1,395 @@
+//! A `DelayQueue` with keyed entries implemented with a `HashMap`.
+//!
+//! A `HashDelayQueue` implements `Stream` which removes expired items from the map.
+
+/// The default delay for entries, in seconds. This is only used when `insert()` is used to add
+/// entries.
+const DEFAULT_DELAY: Duration = Duration::from_secs(30);
+
+use futures::prelude::*;
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio_util::time::delay_queue::{self, DelayQueue};
+
+pub struct HashDelayQueue<K>
+where
+    K: std::cmp::Eq + std::hash::Hash + std::clone::Clone + Unpin,
+{
+    /// The given entries.
+    entries: HashMap<K, delay_queue::Key>,
+    /// A queue holding the timeouts of each entry.
+    expirations: DelayQueue<K>,
+    /// The default expiration timeout of an entry.
+    default_expiration_timeout: Duration,
+}
+
+impl<K> Default for HashDelayQueue<K>
+where
+    K: std::cmp::Eq + std::hash::Hash + std::clone::Clone + Unpin,
+{
+    fn default() -> Self {
+        HashDelayQueue::new(DEFAULT_DELAY)
+    }
+}
+
+impl<K> HashDelayQueue<K>
+where
+    K: std::cmp::Eq + std::hash::Hash + std::clone::Clone + Unpin,
+{
+    /// Creates a new, empty `HashDelayQueue` with a default expiration timeout.
+    pub fn new(default_expiration_timeout: Duration) -> Self {
+        HashDelayQueue {
+            entries: HashMap::new(),
+            expirations: DelayQueue::new(),
+            default_expiration_timeout,
+        }
+    }
+
+    /// Inserts an entry that expires after the default expiration timeout.
+    pub fn insert(&mut self, key: K) {
+        self.insert_with_timeout(key, self.default_expiration_timeout);
+    }
+
+    /// Inserts an entry that will expire after a given duration.
+    ///
+    /// If the key was not present, then a new entry is inserted.
+    /// If the key was present, then the existing entry is updated.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if `timeout` is greater than the maximum duration supported by the
+    /// timer in the current `Runtime`.
+    pub fn insert_with_timeout(&mut self, key: K, timeout: Duration) {
+        if self.contains_key(&key) {
+            // Update the timeout.
+            self.reset_timeout(&key, timeout);
+        } else {
+            let delay_key = self.expirations.insert(key.clone(), timeout);
+            self.entries.insert(key, delay_key);
+        }
+    }
+
+    /// Resets the expiration timeout of an entry for `key`. Returns `true` if the key existed,
+    /// `false` otherwise.
+    ///
+    /// # Panics
```

Comment:
Why panic (here and `insert_with_timeout()`  instead of returning a `Result` of some kind?

Reply:
The panic here is because that is how the [`tokio-util`](https://docs.rs/tokio-util/0.6.8/tokio_util/time/delay_queue/struct.DelayQueue.html#method.insert) library is handling the exception. It looks like the [maximum duration is ~2 years](https://github.com/tokio-rs/tokio/pull/3126), so I think a panic is acceptable here. We should never insert anything with such a long delay.